lib/compress/zstd_compress.c

```diff
@@ -2500,21 +2572,28 @@ static size_t ZSTD_compress_frameChunk (ZSTD_CCtx* cctx,
         /* Ensure hash/chain table insertion resumes no sooner than lowlimit */
         if (ms->nextToUpdate < ms->window.lowLimit) ms->nextToUpdate = ms->window.lowLimit;
 
-        {   size_t cSize = ZSTD_compressBlock_internal(cctx,
-                                op+ZSTD_blockHeaderSize, dstCapacity-ZSTD_blockHeaderSize,
-                                ip, blockSize, 1 /* frame */);
-            FORWARD_IF_ERROR(cSize);
-            if (cSize == 0) {  /* block is not compressible */
-                cSize = ZSTD_noCompressBlock(op, dstCapacity, ip, blockSize, lastBlock);
+        {   int useTargetCBlockSize = ZSTD_useTargetCBlockSize(&cctx->appliedParams);
+            size_t cSize = 0;
+            if (useTargetCBlockSize) {
+                cSize = ZSTD_compressBlock_targetCBlockSize(cctx, op, dstCapacity, ip, blockSize, lastBlock);
                 FORWARD_IF_ERROR(cSize);
             } else {
-                const U32 cBlockHeader = cSize == 1 ?
-                    lastBlock + (((U32)bt_rle)<<1) + (U32)(blockSize << 3) :
-                    lastBlock + (((U32)bt_compressed)<<1) + (U32)(cSize << 3);
-                MEM_writeLE24(op, cBlockHeader);
-                cSize += ZSTD_blockHeaderSize;
+                cSize = ZSTD_compressBlock_internal(cctx,
+                                        op+ZSTD_blockHeaderSize, dstCapacity-ZSTD_blockHeaderSize,
+                                        ip, blockSize, 0 /* frame */);
+                FORWARD_IF_ERROR(cSize);
+
+                if (cSize == 0) {  /* block is not compressible */
+                    cSize = ZSTD_noCompressBlock(op, dstCapacity, ip, blockSize, lastBlock);
+                    FORWARD_IF_ERROR(cSize);
+                } else {
+                    U32 const cBlockHeader24 = lastBlock + (((U32)bt_compressed)<<1) + (U32)(cSize << 3);
```

Comment:
At this line,
the previous version used to differentiate `cSize==1`, to mean "RLE block".
Now the new code only creates `bt_compressed` blocks.

I believe that's because the "RLE block generation" patch has been created after the Superblock one.

Now, it's a mistake, because an RLE block of `cSize==1` will be labelled a compressed block, which is incorrect, and I guess it will become a corruption bug at decompression time.
It seems to prove that we are missing a test able to generate RLE block and check they are correctly encoded and decoded.

Reply:
Yep, I just switched back to the old logic and the tests look good.

Reply:
Just pushed a new commit onto this branch and it doesn't look like Github picks it up - should I open a new PR for the modification?

Reply:
> It seems to prove that we are missing a test able to generate RLE block and check they are correctly encoded and decoded.

The bug is hidden because we've disabled RLE block generation.

Reply:
Yes, a new PR will be needed.